File: problem_solution/lc702_search_sorted_array_unknown_size/SortedArrayUnknownSize.py
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class SortedArrayUnknownSize:
    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        left, right = 0, 1
        
        # Search boundaries
        while (reader.get(right) < target):
            if (reader.get(right) == 2147483647):
                return -1
            left, right = right, right*2 # python int type is unbounded
            
        # Binary search
        while (left <= right):
            mid = left + (right - left)//2
            val = reader.get(mid)
            if val == target:
                return mid
            elif val > target or val == 2147483647:
                right = mid - 1
            else:
                left = mid + 1
        
        return -1

File: problem_solution/lc702_search_sorted_array_unknown_size/SortedArrayUnknownSize.java
/**
 * // This is ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * interface ArrayReader {
 *     public int get(int index) {}
 * }
 */

class SortedArrayUnknownSize {
    public int search(ArrayReader reader, int target) {
        int left = 0;
        int right = 1;
        
        // Search boundaries [Left, Right] to include target
        while ((reader.get(right) < target)) {
            if (reader.get(right) == 2147483647)
                return -1;
            
            left = right;
            if (right < Integer.MAX_VALUE/2)   
                right *= 2;  // or right <<= 1, double right until finding the right boundary (i.e., right > target)
            else{
                right = Integer.MAX_VALUE;
                break;
            }
        }
        // left = right/2;
        
        int mid;
        int val;  // value of reader.get(mid)
        // classic bineary search
        while (left <= right) {
            mid = left + (right - left)/2;
            
            val = reader.get(mid);
            if (val == 2147483647)
                right = mid - 1;
            else if (target < val)
                right = mid - 1;
            else if (target > val)
                left = mid + 1;
            else 
                return mid;
        }
        
        return -1; 
    }
}